dask/blockwise.py

```diff
@@ -30,21 +30,70 @@
 )
 
 
-class BlockwiseIODeps:
-    """Index-argument mapping for Blockwise IO dependencies"""
+class BlockwiseDep:
+    """Base class for (indexable) Blockwise-IO arguments"""
 
-    def __getitem__(self, idx: tuple):
+    numblocks: Tuple[int, ...]
+
+    def __getitem__(self, idx: Tuple[int, ...]) -> Any:
+        raise NotImplementedError(
+            "Must define `__getitem__` for `BlockwiseDep` subclass."
+        )
+
+    def get(self, idx: Tuple[int, ...]) -> Any:
+        return self.__getitem__(idx)
+
+    def get(self, idx: Tuple[int, ...], default) -> Any:
+        try:
+            return self.__getitem__(idx)
+        except KeyError:
+            return default
+
+    def __dask_distributed_pack__(self, output_blocks):
         raise NotImplementedError(
-            "Must define `__getitem__` for `BlockwiseIODeps` subclass."
+            "Must define `__dask_distributed_pack__` for `BlockwiseDep` subclass."
         )
 
     @classmethod
-    def __dask_distributed_pack__(cls, cls_path: str, *args):
-        return (cls_path, *args)
+    def __dask_distributed_pack__(cls, state):
+        raise NotImplementedError(
+            "Must define `__dask_distributed_unpack__` for `BlockwiseDep` subclass."
+        )
+
+    def __repr__(self) -> str:
+        return f"<{type(self).__name__} {self.shape}>"
+
+
+class BlockwiseDepDict(BlockwiseDep):
+    """Base class for dictionary-based Blockwise-IO arguments"""
+
+    def __init__(self, mapping: dict, numblocks: tuple):
+        self.mapping = mapping
+
+        # By default, assume 1D shape
+        self.numblocks = numblocks or (len(mapping),)
+
+    def __getitem__(self, idx: Tuple[int, ...]) -> Any:
+        return self.mapping[idx]
+
+    def __dask_distributed_pack__(self, output_blocks):
+        from distributed.protocol import to_serialize
+
+        packed_mapping = {}
+        for k in output_blocks or self.mapping.keys():
+            v = self.mapping[k]
+            packed_mapping[k] = to_serialize(v)
```

Comment:
> I assume you're doing this for the common case where output_blocks is None

Right.  Unless we have culled, `output_blocks` will typically be `None`.  I totally agree that it would be nice to require that `output_blocks` has already been calculated at this point.  I can try to think through that after everything else feels clean.

Reply:
That makes sense - We definitley want the desired indices for each dependency (rather than the final output indices).  It seems like it will take a bit more work to produce that information at packing time.  Perhaps, for now, we can just design `__dask_distributed_pack__` to accept an optional `required_indices` argument, and not pass anything in for now?

Reply:
Yeah, it's a bit of a refactor to share logic between `make_blockwise_graph` and `__dask_distributed_pack__` for calculating the culled keys.

The optional argument seems good to me. Since this is really just an optimization, another option could be that we simply lie to `__dask_distributed_pack__`, and always pass it the full set of keys (`product(*[range(dim) for dim in arg.numblocks[: len(ind)]] )`), even when culling is in effect. That would give us the final interface we want, just with the wrong implementation. Downside is we do some extra unnecessary work of generating the keys every time.

I have no preference at all between the two; your call.

Reply:
I confirmed that your approach certainly works, and it makes sense to me.  With that said, I'd prefer to leave the argument as "optional" for now, and just pass nothin in.  It still seems possible that we will want to avoid calculating `required_indices` in `Blockwise` whenever we know that there is no culling (`self.output_blocks` is `None`).